packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts

```diff
@@ -195,3 +201,7 @@ export type WebauthnLogin = {
   loginUsernames?: string[];
   onUserResponse?(val: number | string): void;
 };
+
+type LoginParamsWithKind = types.LoginParams & {
+  kind: PrimaryAuthType;
+};
```

Comment:
Ah, it's not quite what I meant. A type like this would still allow some incorrect values to pass through as it says that `LoginParamsWithKind` is any `LoginParams` combined with any `PrimaryAuthType`.

Take a look at this patch:

<details>
<summary>Patch</summary>

```diff
diff --git a/packages/teleterm/src/services/tshd/types.ts b/packages/teleterm/src/services/tshd/types.ts
index 124d4ecb..bc151c9a 100644
--- a/packages/teleterm/src/services/tshd/types.ts
+++ b/packages/teleterm/src/services/tshd/types.ts
@@ -96,28 +96,24 @@ export type TshAbortSignal = {
   removeEventListener(cb: (...args: any[]) => void): void;
 };
 
-export type LoginLocalParams = {
+interface LoginParamsBase {
   clusterUri: string;
+}
+
+export interface LoginLocalParams extends LoginParamsBase {
   username: string;
   password: string;
   token?: string;
-};
+}
 
-export type LoginSsoParams = {
-  clusterUri: string;
+export interface LoginSsoParams extends LoginParamsBase {
   providerType: string;
   providerName: string;
-};
+}
 
-export type LoginPasswordlessParams = {
-  clusterUri: string;
+export interface LoginPasswordlessParams extends LoginParamsBase {
   onPromptCallback(res: WebauthnLoginPrompt): void;
-};
-
-export type LoginParams =
-  | LoginLocalParams
-  | LoginSsoParams
-  | LoginPasswordlessParams;
+}
 
 export type CreateGatewayParams = {
   targetUri: string;
diff --git a/packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts b/packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts
index 03a523c3..4ad48a80 100644
--- a/packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts
+++ b/packages/teleterm/src/ui/ClusterConnect/ClusterLogin/useClusterLogin.ts
@@ -45,27 +45,25 @@ export default function useClusterLogin(props: Props) {
   });
 
   const [loginAttempt, login, setAttempt] = useAsync(
-    (params: LoginParamsWithKind) => {
+    (params: types.LoginParams) => {
       refAbortCtrl.current = clustersService.client.createAbortController();
+
       switch (params.kind) {
         case 'local':
           return clustersService.loginLocal(
-            params as types.LoginLocalParams,
+            params,
             refAbortCtrl.current.signal
           );
         case 'passwordless':
           return clustersService.loginPasswordless(
-            params as types.LoginPasswordlessParams,
+            params,
             refAbortCtrl.current.signal
           );
         case 'sso':
-          return clustersService.loginSso(
-            params as types.LoginSsoParams,
-            refAbortCtrl.current.signal
-          );
+          return clustersService.loginSso(params, refAbortCtrl.current.signal);
         default:
           throw new Error(
-            `loginAttempt: login params kind ${params.kind} not implemented`
+            `loginAttempt: login params kind not implemented: ${params}`
           );
       }
     }
@@ -201,7 +199,3 @@ export type WebauthnLogin = {
   loginUsernames?: string[];
   onUserResponse?(val: number | string): void;
 };
-
-type LoginParamsWithKind = types.LoginParams & {
-  kind: PrimaryAuthType;
-};
diff --git a/packages/teleterm/src/ui/services/clusters/clustersService.test.ts b/packages/teleterm/src/ui/services/clusters/clustersService.test.ts
index f99271c0..ac9e7a05 100644
--- a/packages/teleterm/src/ui/services/clusters/clustersService.test.ts
+++ b/packages/teleterm/src/ui/services/clusters/clustersService.test.ts
@@ -182,6 +182,7 @@ test('login into cluster and sync resources', async () => {
   const client = getClientMocks();
   const service = createService(client, new NotificationsServiceMock());
   const loginParams = {
+    kind: 'local' as const,
     clusterUri,
     username: 'admin',
     password: 'admin',
diff --git a/packages/teleterm/src/ui/services/clusters/types.ts b/packages/teleterm/src/ui/services/clusters/types.ts
index ed3f4897..130557b0 100644
--- a/packages/teleterm/src/ui/services/clusters/types.ts
+++ b/packages/teleterm/src/ui/services/clusters/types.ts
@@ -37,13 +37,18 @@ export type AuthType = shared.AuthType;
 
 export type AuthProvider = tsh.AuthProvider;
 
-export type LoginParams = tsh.LoginParams;
+export type LoginLocalParams = { kind: 'local' } & tsh.LoginLocalParams;
 
-export type LoginLocalParams = tsh.LoginLocalParams;
+export type LoginPasswordlessParams = {
+  kind: 'passwordless';
+} & tsh.LoginPasswordlessParams;
 
-export type LoginPasswordlessParams = tsh.LoginPasswordlessParams;
+export type LoginSsoParams = { kind: 'sso' } & tsh.LoginSsoParams;
 
-export type LoginSsoParams = tsh.LoginSsoParams;
+export type LoginParams =
+  | LoginLocalParams
+  | LoginPasswordlessParams
+  | LoginSsoParams;
 
 export type Application = tsh.Application;
 
```
</details>

This way it's impossible to call `useClusterLogin.login` with `{kind: 'local', providerType: 'foo', providerName: 'bar'}`.

But I see why it might have seemed like I meant doing it the way you did it. When not calling `useClusterLogin.login` but rather one of the underlying methods, such as `ClustersService.loginLocal`, you need to pass an additional field which seems unnecessary (as evidenced by a one line update to a test in that patch). This is just an unfortunate effect of how TypeScript support for sum types looks like.

It works better in situations where we have just a single data structure that's always passed around, like [documents in Connect](https://github.com/gravitational/webapps/blob/bab46b0d3b757c4356a82edc50a11d0510d1bc39/packages/teleterm/src/ui/services/workspacesService/documentsService/types.ts).

Reply:
thank you for the clear explanation!